modules/apps/commerce/commerce-currency-service/src/main/java/com/liferay/commerce/currency/service/impl/CommerceCurrencyLocalServiceImpl.java

```diff
@@ -510,9 +446,22 @@ private void _updateExchangeRates(
 	@ServiceReference(type = ConfigurationProvider.class)
 	private ConfigurationProvider _configurationProvider;
 
+	@ServiceReference(type = CounterLocalService.class)
+	private CounterLocalService _counterLocalService;
+
 	@ServiceReference(type = ExchangeRateProviderRegistry.class)
 	private ExchangeRateProviderRegistry _exchangeRateProviderRegistry;
 
-	private ServiceRegistration<?> _serviceRegistration;
+	@ServiceReference(type = PortalUUID.class)
+	private PortalUUID _portalUUID;
+
+	@ServiceReference(
```

Comment:
This is not going to solve the problem.

We need the CommercePriceListServiceInitialDeploymentUpgradeStepRegistrator(that executes the CommercePriceListServiceVerifyProcess) to wait until the previous initial upgrade process that loads the currencies is executed. That's not easy because at that time the service is already available since it is in the latest schema version.

Let me think about this tomorrow, it might require a change in the upgrade framework.

Reply:
@achaparro We only need to make sure `CommerceCurrencyLocalService` is available after the upgrade step is executed. This is actually a preparation for switching this module from using Spring to DS. I've asked Janis to test her change with the local service using DS, because `@ServiceReference` in spring bean works pretty different with `@Reference`.

But on the other hand, do you mean the `Release` can be available directly with `1.3.0` even before the initial step is executed by "in the latest schema version"?

Reply:
Hi @dantewang, the problem here is that the initialDeploymentUpgradeProcesses are executed once the schema version is in the latest one and the services are available. Those upgrades are intended to load data so using the services is convenient.

However, I came up with one solution, the verify process is executed by this other initial upgrade process:
https://github.com/liferay/liferay-portal/blob/master/modules/apps/commerce/commerce-price-list-service/src/main/java/com/liferay/commerce/price/list/internal/upgrade/registry/CommercePriceListServiceInitialDeploymentUpgradeStepRegistrator.java#L40

We can do the following:
- Disabling it by default:
https://github.com/liferay/liferay-portal/blob/master/modules/apps/commerce/commerce-price-list-service/src/main/java/com/liferay/commerce/price/list/internal/upgrade/registry/CommercePriceListServiceInitialDeploymentUpgradeStepRegistrator.java#L28
- Enabling it when your initial upgrade process is completing. You have an example here:
https://github.com/liferay/liferay-portal/blob/7.4.3.40-ga40/modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/internal/upgrade/registry/ClientExtensionServiceInitialUpgradeStepRegistrator.java#L61

This should work fine because we only need to execute the CommercePriceListServiceInitialDeploymentUpgradeStepRegistrator when the database is new so it can be disabled for the rest of the cases.

What do you think?

cc/ @JanisZhang

Reply:
@achaparro This can solve the possible race condition between these two parts, but there are chances that other classes may also reference `CommerceCurrencyLocalService` and do something once it becomes available. That's why we want the default data to be imported before the service becoming available. If using upgrades, will it work if we register the upgrade process from 0.0.0?

Reply:
Hi @dantewang,

 > If using upgrades, will it work if we register the upgrade process from 0.0.0?
That is not possible because for service modules the 0.0.0 upgrade is reserved for the initial creation of the tables.

Anyway, I agree we need to change the framework. What we can do is to:
1- Implement this for now
2- Create a task so we can change the upgrade framework so the services are not visible to other modules until the initialUpgradeDeployment is executed (it is not trivial because the services should be able to be used there but not outside)

Do you agree?